lib/auth0/mixins/validation.rb

```diff
@@ -85,15 +86,140 @@ def decode_and_validate_signature(id_token, header)
             raise Auth0::InvalidIdToken, 'Invalid ID token signature'
           rescue JWT::IncorrectAlgorithm
             alg = header['alg']
-            raise Auth0::InvalidIdToken, "Signature algorithm of \"#{alg}\" is not supported. Expected the ID token to be signed with \"#{algorithm.name}\""
+            raise Auth0::InvalidIdToken, "Signature algorithm of \"#{alg}\" is not supported. Expected the ID token"\
+                                         " to be signed with \"#{algorithm.name}\""
           rescue JWT::DecodeError
             raise Auth0::InvalidIdToken, "Could not find a public key for Key ID (kid) \"#{kid}\""
           end
         end
-        # rubocop:enable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity
 
-        def validate_claims(decoded_id_token); end # TODO: Implement (in the next PR)
+        # rubocop:disable Metrics/PerceivedComplexity
+        def validate_claims(claims)
+          leeway = @context[:leeway]
+          nonce = @context[:nonce]
+          issuer = @context[:issuer]
+          audience = @context[:audience]
+          max_age = @context[:max_age]
+
+          raise Auth0::InvalidParameter, 'Must supply a valid leeway' unless leeway.is_a?(Integer) && leeway >= 0
+          raise Auth0::InvalidParameter, 'Must supply a valid nonce' unless nonce.nil? || !nonce.to_s.empty?
+          raise Auth0::InvalidParameter, 'Must supply a valid issuer' unless issuer.nil? || !issuer.to_s.empty?
+          raise Auth0::InvalidParameter, 'Must supply a valid audience' unless audience.nil? || !audience.to_s.empty?
+
+          unless max_age.nil? || (max_age.is_a?(Integer) && max_age >= 0)
+            raise Auth0::InvalidParameter, 'Must supply a valid max_age'
+          end
+
+          validate_iss(claims, issuer)
```

Comment:
these validation methods don't need the entire claims map. You could reduce the scope if you pass the value and check for presence comparing against nil, for example.

Reply:
It's done that way to ensure all the logic for verifying a particular claim is inside a single method.

Reply:
> It's done that way to ensure all the logic for verifying a particular claim is inside a single method.

Below the `azp` validation has a few conditions being checked right here

```ruby
validate_azp(claims, audience) if claims['aud'].is_a?(Array) && claims['aud'].count > 1
```

If you continue passing the `claims` to the `validate_azp` method, you could check those conditions inside the method instead, to _really_ keep all the logic for verifying a claim in a single method.

The alternative I was suggesting was passing the value instead of the map.

```ruby
validate_azp(claims['azp'], audience) if claims['aud'].is_a?(Array) && claims['aud'].count > 1
```

As it is today, it is a mix of both.

Reply:
Those bits of conditional logic are checking preconditions. The claims validations logic is still all in the method.